src/cr/cube/dimension.py

```diff
@@ -610,6 +612,50 @@ def _dimension_transforms_dict(self) -> Dict:
         """Copy of dimension transforms dictionary with shimmed `element_id`s"""
         return self._element_id_shim.shimmed_dimension_transforms_dict
 
+    @lazyproperty
+    def _element_data_format(self) -> Optional[Union[str, int]]:
+        """optional str format for datetimes or int number of decimals for numeric"""
+        return self._dimension_dict["references"].get("format", {}).get("data")
+
+    def _format_datetime_element_label(self, x: Any) -> str:
+        """returns str of x formatted for use in labels with desired date format"""
+        orig_format = self._incoming_element_datetime_format
+        out_format = self._element_data_format
+        if orig_format is None or out_format is None:
+            return str(x)
+        try:
+            return datetime.strptime(x, orig_format).strftime(out_format)
+        except ValueError:
+            return str(x)
+
+    def _format_element_label(self, x: Any) -> str:
+        """returns str of x formatted for use in labels"""
+        if self.dimension_type == DT.DATETIME:
+            return self._format_datetime_element_label(x)
+        else:  # --- TODO: Should round numeric values like the frontend does
+            return str(x)
+
+    @lazyproperty
+    def _incoming_element_datetime_format(self) -> Optional[str]:
+        """optional str of date formatting requested if datetime and is found"""
+        if self.dimension_type != DT.DATETIME:
+            return None
```

Comment:
is there any reason why all this methods/properies belong to the `_Dimension` class? I'm for having this in the `_AllElements` maybe or in the `_Element` itself. Probably passing the `_element_data_format` to the `_AllElements` is all you need to move all this logic there. Is there any impediment that I don't see? What do you think?

Reply:
Yeah, originally wanted it in `_Element`, but then I'd have to pass `dim_type`, `incoming_format` and `desired_format` into each element. Maybe in `_AllElements` is better because the only thing new would be `incoming_format`